cache/refs.go

```diff
@@ -739,15 +744,24 @@ func (sr *immutableRef) getBlobWithCompression(ctx context.Context, compressionT
 	if err != nil {
 		return ocispecs.Descriptor{}, err
 	}
-	return getBlobWithCompression(ctx, sr.cm.ContentStore, desc, compressionType)
+	return sr.getBlobWithCompressionWithDesc(ctx, sr.cm.ContentStore, desc, compressionType)
 }
 
-func getBlobWithCompression(ctx context.Context, cs content.Store, desc ocispecs.Descriptor, compressionType compression.Type) (ocispecs.Descriptor, error) {
+func (sr *immutableRef) getBlobWithCompressionWithDesc(ctx context.Context, cs content.Store, blobDesc ocispecs.Descriptor, compressionType compression.Type) (ocispecs.Descriptor, error) {
 	if compressionType == compression.UnknownCompression {
 		return ocispecs.Descriptor{}, fmt.Errorf("cannot get unknown compression type")
 	}
 	var target *ocispecs.Descriptor
-	if err := walkBlob(ctx, cs, desc, func(desc ocispecs.Descriptor) bool {
+	if err := walkBlob(ctx, cs, blobDesc, func(key string, desc ocispecs.Descriptor) bool {
+		if isNydusCompressionType(compressionType) && compressionType == getNydusCompressionType(desc) {
+			// Maybe we can find multiple nydus bootstraps/blobs that are linked to
+			// blobDesc, the key check ensures to get unique nydus bootstrap/blob.
```

Comment:
Shouldn't they be deduplicated?

Reply:
> Now we seem to have 2 implementation of `computeBlobChain`(`computeBlobChain` and `computeNydusBlobChain`). Some concerns about maintainability:
> 
> * How can guarantee/keep the compatibility of these implementations?
> * Will we need 2 duplicated patches on these implementations for 1 futural change?

@ktock Thanks for the reply. nydus is using another binary tool named `nydus-image` to implement blob diff computation by self between a lower snapshot and upper snapshot, and generate these two kinds of artifacts:

- nydus blob: the file data is added and modified on the upper snapshot, no deleted data.
- nydus bootstrap: the file metadata on the whole view of the upper snapshot.

So it's hard to re-use `computeBlobChain`, especially for `doCompression` (be introduced in the [commit](https://github.com/moby/buildkit/pull/2581/commits/3cb0a7026a7613a56591ab8f64b013658b1319ad)).

Is it possible to use `sr.layerWalk` in `computeBlobChain` to re-use part of codes?

Reply:
> Shouldn't they be deduplicated?

For handling the case if have different nydus bootstraps are linked to a gzip blob:

```
// Create gzip compression blob
b, gzipDesc1, err := mapToBlob(map[string]string{"gzip1": "true"}, true)
err = content.WriteBlob(ctx, co.cs, "ref-gzip-1", bytes.NewBuffer(b), gzipDesc1)

// Create ref 1
active, err := cm.New(ctx, nil, nil)
snap, err := active.Commit(ctx)
snapRef1 := snap.(*immutableRef)
// Set gzip compression to ref 1
err = snapRef1.setBlob(ctx, gzipDesc1)

// Create ref 2
active, err = cm.New(ctx, nil, nil)
snap, err = active.Commit(ctx)
snapRef2 := snap.(*immutableRef)
// Set gzip compression to ref 2
err = snapRef2.setBlob(ctx, gzipDesc1)

// Link nydus bootstrap compression 1 to ref 1
b, bootstrapDesc1, err := mapToNydusArtifact("nydus-bootstrap-1", compression.NydusBootstrap)
err = content.WriteBlob(ctx, co.cs, "ref-nydus-bootstrap-1", bytes.NewBuffer(b), bootstrapDesc1)
err = snapRef1.SetString(keyCachedNydusBootstrap, bootstrapDesc1.Digest.String(), "")
err = snapRef1.linkBlob(ctx, bootstrapDesc1)

// Link nydus bootstrap compression 2 to ref 2
b, bootstrapDesc2, err := mapToNydusArtifact("nydus-bootstrap-2", compression.NydusBootstrap)
err = content.WriteBlob(ctx, co.cs, "ref-nydus-bootstrap-2", bytes.NewBuffer(b), bootstrapDesc2)
err = snapRef2.SetString(keyCachedNydusBootstrap, bootstrapDesc2.Digest.String(), "")
err = snapRef2.linkBlob(ctx, bootstrapDesc2)

// Get nydus bootstrap compression from ref 1
actualDesc, err = snapRef1.getBlobWithCompression(ctx, compression.NydusBootstrap)
// Should get the bootstrap linked to ref 1
require.Equal(t, bootstrapDesc1, actualDesc)

// Get nydus bootstrap compression from ref 2
actualDesc, err = snapRef2.getBlobWithCompression(ctx, compression.NydusBootstrap)
// Should get the bootstrap linked to ref 2
require.Equal(t, bootstrapDesc2, actualDesc)
```

Reply:
@ktock @sipsma Updated: Indeed the current implementation has a bit dirty, I’m working on a new solution to reuse the current computeBlobChain method, will re-commit later, thanks for following the PR!